app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt

```diff
@@ -0,0 +1,135 @@
+import java.math.BigInteger
+
+class AssetTransferViewHolder(
+    private val safe: Solidity.Address,
+    private val data: TransactionData.AssetTransfer,
+    private val addressHelper: AddressHelper,
+    private val tokenRepository: TokenRepository
+) : TransactionInfoViewHolder {
+
+    private val disposables = CompositeDisposable()
+    private var view: View? = null
+
+    override fun loadTransaction(): Single<SafeTransaction> =
+        Single.fromCallable {
+            AssetTransferTransactionBuilder.build(data)
+        }.subscribeOn(Schedulers.computation())
+
+    override fun inflate(inflater: LayoutInflater, root: ViewGroup) {
+        view = inflater.inflate(R.layout.layout_asset_transfer_info, root, true)
+    }
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_START)
+    fun start() {
+        val view = view ?: return
+        view.layout_asset_transfer_info_value.text = "~"
+        view.layout_asset_transfer_info_fiat.text = "~ $"
+
+        setupSafeInfo()
+        setupToInfo()
+        setupTokenInfo()
+    }
+
+    private fun setupTokenInfo() {
+        val view = view ?: return
+        view.layout_asset_transfer_info_value.text = "~ $"
+        disposables += loadTokenInfo()
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeBy(
+                onSuccess = {
+                    view.layout_asset_transfer_info_value.text = it.displayString(data.amount)
+                    loadSafeTokenBalance(it)
+                },
+                onError = {
+                    view.layout_asset_transfer_info_value.text = data.amount.asDecimalString()
+                    view.layout_asset_transfer_info_fiat.text = view.context.getString(R.string.unknown_token)
+                    loadSafeTokenBalance()
+                }
+            )
+    }
+
+    private fun loadSafeTokenBalance(token: ERC20Token? = null) {
+        val view = view ?: return
+        val erc20Token = token ?: ERC20Token(data.token, decimals = 0)
+        // We build a erc20 token wrapper to request the balance
+        disposables += tokenRepository.loadTokenBalances(safe, listOf(erc20Token))
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeBy(onNext = {
+                val balance = it.firstOrNull()?.second ?: BigInteger.ZERO
+                view.layout_asset_transfer_info_safe_balance.text = token?.displayString(balance) ?: balance.asDecimalString()
+            }, onError = {
+                view.layout_asset_transfer_info_safe_balance.text = null
+            })
+    }
+
+    private fun loadTokenInfo() =
+        if (data.token == ERC20Token.ETHER_TOKEN.address)
+            Single.just(ERC20Token.ETHER_TOKEN)
+        else
+            tokenRepository.loadToken(data.token)
+                .onErrorResumeNext {
+                    tokenRepository.loadTokenInfo(data.token).firstOrError()
+                }
+
+    private fun setupSafeInfo() {
+        val view = view ?: return
```

Comment:
These checks on every method are not the best 🤔

Reply:
why "not the best" ... they don't look nice I agree. I could also use lateinit, but that would potentially crash the app. I will change some of them to use `apply {}` instead ... looks nicer